### PR TITLE
#2221 - plugin.js update for sub-counter fix

### PR DESCRIPTION
**Pull request by shermandoug**
see issue:
CityofOttawa/ottawa_profile#2221

Adjusted plugin.js css for dub-item counter removal

### CODE REVIEW

File: plugin.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
...
+        'ol.bylawlist,ol.bylawlist>li,ol.bylawlist>li ol>li,ol.bylawlist>li ol>li ol>li, ol.bylawlist>li ol>li ol>li ol>li{list-style-type:none;list-style-position:inside}ol.bylawlist{padding-left:1.5em}ol.bylawlist>li{counter-increment:first}ol.bylawlist>li:before{content:"(" counter(first,decimal) ") "}ol.bylawlist>li ol>li{counter-increment:second}ol.bylawlist>li ol>li:before{content:"(" counter(second,lower-alpha) ") "}ol.bylawlist>li ol>li ol>li{counter-increment:third}ol.bylawlist>li ol>li ol>li:before{content:"(" counter(third,lower-roman) ") "}ol.bylawlist>li ol>li ol>li ol>li{counter-increment:fourth}ol.bylawlist>li ol>li ol>li ol>li:before{content:"(" counter(fourth, decimal) ") "}'
```
Comment (minorOffense) [inline plugin.js]: There's still brackets at the fourth level. Remove them so it outputs ``1.`` ``2.`` ``3.`` etc.

**Review by minorOffense — CHANGES_REQUESTED** [review]

File: plugin.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
+        'ol.bylawlist,ol.bylawlist>li,ol.bylawlist>li ol>li,ol.bylawlist>li ol>li ol>li, ol.bylawlist>li ol>li ol>li ol>li{list-style-type:none;list-style-position:inside}ol.bylawlist{padding-left:1.5em}ol.bylawlist>li{counter-increment:first}ol.bylawlist>li:before{content:"(" counter(first,decimal) ") "}ol.bylawlist>li ol>li{counter-increment:second}ol.bylawlist>li ol>li:before{content:"(" counter(second,lower-alpha) ") "}ol.bylawlist>li ol>li ol>li{counter-increment:third}ol.bylawlist>li ol>li ol>li:before{content:"(" counter(third,lower-roman) ") "}ol.bylawlist>li ol>li ol>li ol>li{counter-increment:fourth}ol.bylawlist>li ol>li ol>li ol>li:before{content: counter(fourth, decimal)}'
```
Comment (elcazador) [inline plugin.js]: I thought we talked about leaving the space after the number

File: plugin.js [hunk 3]
```diff
@@ -25,7 +25,7 @@ CKEDITOR.plugins.add( 'bylawlist', {
         editor.ui.addButton( 'Bylawlist', {
             label: 'Add Bylawlist',
             command: 'insertBylawlist',
-            toolbar: 'insert'g
+            toolbar: 'insert'
```
Comment (elcazador) [inline plugin.js]: Why is this needed?  Master doesn't have the typo.  Did you miss something

**Review by elcazador — CHANGES_REQUESTED** [review]

**Review by minorOffense — APPROVED** [review]

**Comment by minorOffense** [comment]
This is updated with the new JS and the new CSS to show the custom numbers and the fourth level with the ``1.`` pattern.

File: plugin.js [hunk 4]
```diff
@@ -1,36 +1,70 @@
...
-                  else {
-                    bylaw.className = "";
-                  }
-                  
-                }              
-              }
+CKEDITOR.plugins.add('bylawlist', {
+  onLoad: function () {
+    CKEDITOR.addCss(
+      'ol.bylawlist{list-style-type:none;list-style-position:inside;padding-left:1.5em}ol.bylawlist>li>ol,ol.bylawlist>li>ol>li>ol{list-style-type:inherit}ol.bylawlist li{counter-increment:bylawlist-counter}ol.bylawlist li:first-child{counter-reset:bylawlist-counter}ol.bylawlist>li:before{content:"(" counter(bylawlist-counter,decimal) ") "}ol.bylawlist>li>ol>li:before{content:"(" counter(bylawlist-counter,lower-alpha) ") "}ol.bylawlist>li>ol>li>ol>li:before{content:"(" counter(bylawlist-counter,lower-roman) ") "}ol.bylawlist>li>ol>li>ol>li>ol{list-style-type:decimal}ol.bylawlist>li>ol>li>ol>li>ol>li:before{content:none}'
```
Comment (minorOffense) [inline plugin.js:4]: This is the compiled copy from the theme for the fourth level bylawlist

File: plugin.js [hunk 5]
```diff
@@ -1,36 +1,70 @@
...
+    var format = {
+      'element': 'Bylawlist'
+    };
+
+    var style = new CKEDITOR.style(format);
+
+    editor.attachStyleStateChange(style, function (state) {
+      !editor.readOnly && editor.getCommand('insertBylawlist').setState(state);
+    });
+    editor.on('contentDom', function () {
```
Comment (minorOffense) [inline plugin.js:41]: This is where most of the complexity was. Since the ckeditor uses an iframe / custom document element we had to jump into that document from the editor to attach the observers. This is what makes the numbers change when a user sets the starting value. This only works in browsers that support Mutation observers but Vennila said that's fine since all their authors are told to use Chrome.

File: plugin.js (hunk 4 above)
Comment (walkerjam) [reply plugin.js:4]: Gross

Comment (minorOffense) [reply plugin.js:4]: It’s just the css from _list.scss in the theme. So the lists show up in the editor as they would on display. You don’t actually edit this you just copy / paste the compiled version from Ottawa bootstrap.

File: plugin.js [hunk 6]
```diff
@@ -1,36 +1,70 @@
...
+      const options = {
+        attributeFilter: ['start'],
+        subtree: true
+      };
+
+      const observer = new MutationObserver(function (mutations) {
+        mutations.forEach(function (mutation) {
+          updateList(mutation.target);
         });
-        var format = {'element': 'Bylawlist'};
```
Comment (walkerjam) [inline plugin.js:30]: This block no longer needed?

Comment (minorOffense) [reply plugin.js:30]: No it just moved. See line 31 or so.

Comment (walkerjam) [reply plugin.js:30]: Gotcha!

File: plugin.js [hunk 7]
```diff
@@ -1,36 +1,70 @@
...
-        var style = new CKEDITOR.style(format);
-        editor.attachStyleStateChange( style, function( state ) {
-   	      !editor.readOnly && editor.getCommand( 'insertBylawlist' ).setState( state );
-        } );
-    }
+      });
+
+      observer.observe(target, options);
```
Comment (walkerjam) [inline plugin.js:69]: Can we reduce the scope of the target to just the ckeditor contents?

Comment (minorOffense) [reply plugin.js:69]: It already does. The obj is the body of the ckeditor text area. There’s a call up above for it.

Comment (walkerjam) [reply plugin.js:69]: Ok thanks, wasn't sure what `editor.document` refers to. Maybe a comment there would be helpful for later.

**Review by walkerjam — APPROVED** [review]

**Review by elcazador — APPROVED** [review]
